README.md
# FEmailAll

EmailAll 的 Bug fixed 版本，主要修复了 phonebook、snov、veryvp 模块无法收集邮箱的bug。

## 说明

感谢作者，指路原项目 https://github.com/Taonn/EmailAll

挺好的一个收集邮箱的工具，但由于有些功能无法工作，原作者不再维护，觉得有些可惜，所以简单修补了一下。

如果想要验证邮箱的有效性可以尝试我的另一个项目 https://github.com/ccc-f/Femail

## 0x2 安装&使用

```bash
$ git clone https://github.com/ccc-f/FEmailAll.git
$ cd EmailAll
$ pip3 install -r requirements.txt
```

```python
EmailAll is a powerful Email Collect tool

Example:
    python3 emailall.py check
    python3 emailall.py --domain example.com run
    python3 emailall.py --domains ./domains.txt run
```

` python3 emailall.py --domain example.com run`

`python3 emailall.py --domains ./domains.txt run`

最终结果保存至`result`目录，分为不同模块json数据文件和汇总`{domain}_All.json` 和 `{domain}_All.txt` 文件。

## 模块

邮箱信息收集主要来源如下：

 ```
 - Search
  - Ask                     	# 需代理
  - Baidu                   	# 反爬虫，无法正常使用
  - Bing                    	# cn 版可以，国际版需代理
  - Google                  	# 反爬虫，无法正常使用，加需代理
  - QWant                   	# 正常
  - SO                      	# 反爬虫，无法正常使用
  - Sougou                  	# 反爬虫，无法正常使用
  - GithubApi               	# 正常
- DataSets
  - Email-Format            	# 正常
  - Skymem                  	# 正常
  - Veryvp                  	# 正常，已修复
  - PhoneBook               	# 正常，已修复
  - Snov                    	# 正常，已修复
 ```

​    

## 配置

proxy配置在 setting.py文件

proxy={'http': '127.0.0.1:2333', 'https': '127.0.0.1:2333'}

API配置在 api.py文件

```
# http://www.veryvp.com/

veryvp_username = ''
veryvp_password = ''

# https://www.github.com

github_token = ''

# https://app.snov.io/

snov_username = ''
snov_password = ''

# https://phonebook.cz/

pb_url = 'https://2.intelx.io/phonebook/search'
pb_key = ''

- 
```

`veryvp`和`snov`去网站免费注册

- `GitHub`的token去设置里创建一个即可

phonebook 的 apikey 的获取方式更新了，先打开 https://phonebook.cz/ ，按提示注册登录，再返回原页面，按F12，搜索框搜索任意一个域名，可以在 Network 里面看到POST请求 	https://2.intelx.io/phonebook/search?k=xxx-xxx-xxx-xxx ， 把key放到配置文件里面的 pb_key 就可以了。

## 免责声明

​	本工具仅能在取得足够合法授权的企业安全建设中使用，在使用本工具过程中，您应确保自己所有行为符合当地的法律法规。 如您在使用本工具的过程中存在任何非法行为，您将自行承担所有后果，本工具所有开发者和所有贡献者不承担任何法律及连带责任。 除非您已充分阅读、完全理解并接受本协议所有条款，否则，请您不要安装并使用本工具。 您的使用行为或者您以其他任何明示或者默示方式表示接受本协议的，即视为您已阅读并同意本协议的约束。


docs/changes.md
## 2022.12.06

修复一些bug，by ccc-f。

## 2022.2.19

- 新增PhoneBook、Snov搜集模块，数据结果更全！！！

![PhoneBook&Snov_result](../img/PhoneBook&Snov.png)

![PhoneBook&Snov_result](../img/PhoneBook&Snov_result.png)

## 2022.2.16

- 新增Github搜集模块，更全！！！

![githubapi.png](../img/githubapi.png)

## 2022.2.14

- [EmaiAll](https://github.com/Taonn/EmailAll)发布

common/utils.py
#!/usr/bin/python3
# coding=utf-8
# Author: @Tao.
import random
import requests
import re
import json

from config.log import logger
from pathlib import Path
from config import settings


def check_net():
    """
    检查网络
    :return:
    """
    urls = ['http://ip-api.com/json/']
    url = random.choice(urls)
    header = {'User_Agent': 'curl'}
    timeout = 5
    verify = False
    logger.log('DEBUG', f'Trying to access {url}')
    session = requests.Session()
    session.trust_env = False
    try:
        rsp = session.get(url, headers=header, timeout=timeout, verify=verify)
    except Exception as e:
        logger.log('ERROR', e.args)
        logger.log('ALERT', 'Unable to access Internet, retrying...')
        raise e
    logger.log('DEBUG', 'Access to Internet OK')
    country = rsp.json().get('country').lower()
    if country in ['cn', 'china']:
        logger.log('DEBUG', f'The computer is located in China')
        return True, True
    else:
        logger.log('DEBUG', f'The computer is not located in China')
        return True, False


def get_net_env():
    logger.log('INFOR', 'Checking network environment')
    try:
        result = check_net()
    except Exception as e:
        logger.log('DEBUG', e.args)
        logger.log('ALERT', 'Please check your network environment.')
        return False, None
    return result


def read_file(path):
    """
    读取domains传入的文件内容
    :param path: domains文件路劲
    :return:
    """
    domains = list()
    with open(path, encoding='utf-8', errors='ignore') as f:
        for line in f:
            domain = line.strip('\r\n')
            if not domain:
                continue
            domains.append(domain)
        sorted_domains = sorted(set(domains), key=domains.index)

        return sorted_domains


def get_from_domain(domain):
    domains = set()
    if isinstance(domain, str):
        if domain.endswith('.txt'):
            logger.log('FATAL', 'Use domains parameter')
            exit(1)
        domain = domain.lower().strip()
        if not domain:
            return domains
        domains.add(domain)
    return domains


def get_from_domains(domains):
    domains_data = set()
    if not isinstance(domains, str):
        return domains_data
    try:
        path = Path(domains)
    except Exception as e:
        logger.log('ERROR', e.args)
        return domains_data
    if path.exists() and path.is_file():
        domains = read_file(domains)
        return domains
    return domains_data


def get_domains(domain, domains=None):
    logger.log('DEBUG', f'Getting domains')
    domain_data = get_from_domain(domain)
    domains_data = get_from_domains(domains)
    final_domains = list(domain_data.union(domains_data))
    if domains_data:
        final_domains = sorted(final_domains, key=domains_data.index)
    if not final_domains:
        logger.log('ERROR', f'Did not get a valid domain')
    logger.log('DEBUG', f'The obtained domains \n{final_domains}')

    return final_domains

def check_response(method, resp):
    """
    检查响应 输出非正常响应返回json的信息

    :param method: 请求方法
    :param resp: 响应体
    :return: 是否正常响应
    """
    if resp.status_code == 200 and resp.content:
        return True
    logger.log('ALERT', f'{method} {resp.url} {resp.status_code} - '
                        f'{resp.reason} {len(resp.content)}')
    content_type = resp.headers.get('Content-Type')
    if content_type and 'json' in content_type and resp.content:
        try:
            msg = resp.json()
        except Exception as e:
            logger.log('DEBUG', e.args)
        else:
            logger.log('ALERT', msg)
    return False


def match_emails(domain, html):
    """
    匹配邮箱
    :param domain:要配备邮箱的域名
    :param html: 爬取的网页
    :return: 网页解析的邮箱合集
    """
    # reg_emails = re.compile(r'[a-zA-Z0-9.\-_+#~!$&\',;=:]+' + '@' + '[a-zA-Z0-9.-]*' + domain.replace('www.', ''))
    reg_emails = re.compile(r'[a-zA-Z0-9.\-_]+' + '@' + '[a-zA-Z0-9.-]*' + domain.replace('www.', ''))
    temp = reg_emails.findall(html)
    emails = list(set(temp))
    true_emails = {str(email)[1:].lower().strip().replace('mailto:', '') if len(str(email)) > 1 and str(email)[0] == '.'
                   else len(str(email)) > 1 and str(email).lower().strip() for email in emails}
    return true_emails


def save_all(domain):
    """
    汇总 所有获取域名的邮箱保存至汇总json和txt文件
    :param domain:
    :return:
    """
    logger.log('TRACE', f'Save the Email results found by '
                        f'All module as a json file')
    path = settings.result_save_dir.joinpath(domain.replace('.', '_'))
    filename = domain + '_All' + '.json'
    txt_filename = domain + '_All' + '.txt'
    json_path = path.joinpath(filename)
    txt_path = path.joinpath(txt_filename)
    emails = set()
    for datas in settings.emails:
        for data in datas:
            emails.update(data['emails'])
    with open(txt_path, 'a', encoding='utf-8')as f:
        for email in emails:
            f.write(email + '\n') 

    results = {
        'total': len(emails),
        'emails': list(emails)}
    with open(json_path, 'w', errors='ignore') as f:
        json.dump(results, f, ensure_ascii=False, indent=4)
    logger.log('TRACE', f'Save the ALl Email results found by '
                        f'All module as a file')

modules/datasets/veryvp.py
#!/usr/bin/python3
# -*- coding:utf-8 -*- 
#
# @name   : EmailAll - Email Information Gathering Tools
# @url    : http://github.com/Taonn
# @author : Tao. (Taonn)
# bug fixed by ccc-f 2022/12/06
from common.search import Search
from urllib.parse import unquote,quote
from config import settings
import requests

import json


class Veryvp(Search):
    def __init__(self, domain):
        Search.__init__(self)
        self.domain = domain
        self.module = 'Datasets'
        self.source = 'Veryvp'
        self.session = requests.session()
        self.addr = 'http://www.veryvp.com/SearchEmail/GetEmailList'
        self.num = 999

    def search(self):
        login_url = "http://veryvp.com/user/Login"
        login_par = {"UserName":settings.veryvp_username,
                     "Password":settings.veryvp_password,
                     "ValidateCode":"",
                     "KeepPassword":"0"}
        login_par = {
            'json': unquote(str(login_par)),
        }
        
        rep = self.session.get(login_url)
        rep = self.session.post(login_url,login_par)
        if "登录成功" in rep.text:
            self.cookie = requests.utils.dict_from_cookiejar(rep.cookies)

            params = {'Key': self.domain,
                      'Order': '',
                      'PageSize': self.num,
                      'PageNo': 1}
            self.header.update({'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'})
            resp = self.session.post(url=self.addr, data=params)
            result = json.loads(resp.text)
            data = eval(result)
            data = data['Data']
            data = unquote(data)
            emails = list()

            if isinstance(data,str):
                for e in eval(data):
                    emails.append(e["email"])
                self.results.update(emails)
            else:
                pass
        else:
            pass

    def run(self):
        self.begin()
        self.search()
        self.finish()
        self.save_json()
        self.save_res()


def run(domain):
    search = Veryvp(domain)
    search.run()


if __name__ == '__main__':
    run('example.com')


modules/datasets/phonebook.py
#!/usr/bin/python3
# -*- coding:utf-8 -*-
#
# @name   : EmailAll - Email Information Gathering Tools
# @url    : http://github.com/Taonn
# @author : Tao. (Taonn)
# bug fixed by ccc-f 2022/12/06
from common.search import Search
from config import settings
from config.log import logger


class PhoneBook(Search):
    def __init__(self, domain):
        Search.__init__(self)
        self.domain = domain
        self.module = 'Datasets'
        self.source = 'PhoneBook'
        self.addr = settings.pb_url
        self.num = 999
        self.key = settings.pb_key if settings.pb_key else "xxxxx"

    def search(self):

        self.header.update({'Referer': 'https://phonebook.cz/',
                            'Origin': 'https://phonebook.cz',
                            'sec-ch-ua-platform': '"Windows"',
                            'Sec-Fetch-Site': 'cross-site',
                            'Sec-Fetch-Mode': 'cors',
                            'Sec-Fetch-Dest': 'empty',
                            'sec-ch-ua': '" Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"',
                            'Accept': '* / *',
                            'Content-Type': 'application / json',
                            })
        params = {"term": self.domain,
                  "maxresults": 10000,
                  "media": 0,
                  "target": 2,
                  "terminate": ["{}".format(settings.pb_key)],
                  "timeout": 20}
        url = self.addr + f'?k={self.key}'
        resp = self.post(url, data=params)
        if not resp:
            return
        if hasattr(resp, 'json'):
            id = resp.json()["id"]
            if not id:
                logger.log('ALERT', f'Get PhoneBook id fail')
                return
            params = {
                "k": self.key,
                "id": id,
                "limit": 10000
            }
            self.header.update({'TE': 'trailers'})
            rep = self.get(self.addr + '/result', params=params)
            if hasattr(rep, 'json'):
                emails = self.match_emails(rep)
                if emails:
                    self.results.update(emails)
                else:
                    pass

    def run(self):
        self.begin()
        self.search()
        self.finish()
        self.save_json()
        self.save_res()


def run(domain):
    search = PhoneBook(domain)
    search.run()


if __name__ == '__main__':
    run('example.com')


modules/datasets/snov.py
#!/usr/bin/python3
# -*- coding:utf-8 -*-
#
# @name   : EmailAll - Email Information Gathering Tools
# @url    : http://github.com/Taonn
# @author : Tao. (Taonn)
# bug fixed by ccc-f 2022/12/06
import requests
import json
import re

from common.search import Search
from urllib.parse import unquote, quote
from config import settings
from config.log import logger


class Snov(Search):
    def __init__(self, domain):
        Search.__init__(self)
        self.domain = domain
        self.module = 'Datasets'
        self.source = 'Snov'
        self.addr = 'https://app.snov.io'
        self.username = settings.snov_username
        self.password = settings.snov_password
        self.num = 999
        self.session = requests.session()

    def login(self):

        resp = self.session.get(self.addr + '/login')
        if resp:
            token = re.findall(r'"csrf-token" content="(.*)"', resp.text)[0]
            param = {
                "email": settings.snov_username,
                "password": settings.snov_password,
                "remember": True
            }
            self.header.update({
                        'X-Csrf-Token': token,
                        'X-Requested-With': 'XMLHttpRequest',
                        'Content-Type': 'application/json;charset=utf-8',
                        'Referer': 'https://app.snov.io/login',
                        'Origin': 'https://app.snov.io'
                    })
            if token:
                rep = self.session.post(self.addr + '/login', headers=self.header, json=param)
                if rep.status_code == '301':
                    return True
            else:
                return

    def search(self):

        if self.login():
            resp = self.session.get(self.addr + '/domain-search')
            if resp:
                token = re.findall(r'"_meta" content="(.*)"', resp.text)[0]
                self.header.update({
                    'X-Csrf-Token': token,
                    })
            self.header.update({'Content-Type': 'application/json;charset=UTF-8'})
            params = {"domain": self.domain,
                      "isGreen": 'true',
                      "lastId": 0,
                      "perPage": 20}
            resp = self.session.post(self.addr + '/domain-search', headers=self.header, json=params)
            if hasattr(resp, 'json'):
                resp_json = resp.json()
                cid = resp_json["companyInfo"]["id"]
                params = {"isGreen": True,
                          "lastId": 0,
                          "perPage": self.num,
                          "page": 1}
                rep = self.session.post(self.addr + f'/companies/{cid}', headers=self.header)
                self.header.update({'Referer': f'https://app.snov.io/domain-search?name={self.domain}&tab=emails',
                                    'X-Requested-With': 'XMLHttpRequest'})
                rep = self.session.post(self.addr + f'/companies/{cid}/emails', headers=self.header, json=params)
                emails = self.match_emails(rep)
                if emails:
                    self.results.update(emails)
                else:
                    return
        else:
            logger.log('ALERT', 'Search Failed! you need to configure the username&password in api.py file')
            return

    def run(self):
        self.begin()
        self.search()
        self.finish()
        self.save_json()
        self.save_res()


def run(domain):
    search = Snov(domain)
    search.run()


if __name__ == '__main__':
    run('pingan.com.cn')
